### DISCUSSION TITLE
How to triangulate points in 3d with triangles (not tetrahedrons) using CGAL?

### DISCUSSION THREAD

**Discussion by MartensCedric** [CGAL/cgal#7716] [Q&A] [answered]
[Question on Stack Overflow](https://stackoverflow.com/questions/77115339/how-to-triangulate-points-in-3d-with-triangles-not-tetrahedrons-using-cgal)

Using CGAL, I have some points in R^3, they represent a 2-manifold that is closed and embedded in R^3. For the purpose of illustration, I have a S^2 sphere:

```
for(int i = 0; i < num_vertices; i++)
{
    double theta = 2 * CGAL_PI * i / num_vertices;
    double phi = CGAL_PI * i / num_vertices;
    double x = std::sin(phi) * std::cos(theta);
    double y = std::sin(phi) * std::sin(theta);
    double z = std::cos(phi);
    points.emplace_back(x,y,z);
}
```
I would like to triangulate this sphere using CGAL and obtain a triangle mesh. It seems to me that Triangulation_3 generates tetrahedral meshes. I would like each facet to have 3 vertices, not 4. In the [documentation](https://doc.cgal.org/latest/Triangulation_3/index.html), it does seem that degenerate cases are supported, but I'm struggling to find how to properly use triangle connectivity.

**Comment by stla** [comment]
Hello,

A tetrahedron is made of triangles so what do you have against the tetrahedra? If you have tetrahedra then you have triangles. Do you mean you don't want some edges that go through the interior of your manifold ?

**Reply by MartensCedric** [reply]
Yes, no edges that go through the interior

**Reply by stla** [reply]
You won't see them if you color the mesh ^^ No, seriously I don't know.

**Reply by MartensCedric** [reply]
Unfortunately, I care more about the topology more than looking at it

**Reply by stla** [reply]
[Here](https://stackoverflow.com/questions/996620/how-to-determine-if-a-delaunay-triangle-is-internal-or-external) you may find some ideas This is for 2D and the question is the opposite of yours but maybe that can be generalizable, I don't know.

**Reply by stla** [reply]
How does the shape of your manifold look like? Is it convex? What could you do with a torus? Maybe the alpha hull or the alpha wrapping could give something.

**Reply by MartensCedric** [reply]
it's convex, it's points on a sphere (not like I showed in the code) but I want to mesh the sphere using these points as vertices

**Reply by sloriot** [reply, marked as answer]
https://doc.cgal.org/latest/Convex_hull_3/Convex_hull_3_2quickhull_3_8cpp-example.html

**Reply by afabri** [reply]
Note that you wrote "For the purpose of illustration, I have a S^2 sphere:"  , so we all assumed that you are after the more general case "points in R^3, they represent a 2-manifold that is closed and embedded in R^3"   for which the convex hull algorithm is NOT the solution.

**Reply by MartensCedric** [reply]
Yep, my bad. I said that because I didn't want people to give my sphere meshing algorithms, and I thought my case was general until I realized my case was convex!

**Comment by afabri** [comment]
In fact you should have a look at the surface reconstuction algorithms. To get started have a look at the [tutorial](https://doc.cgal.org/latest/Manual/tuto_reconstruction.html).
